lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,155 @@
+require 'puppetserver/ca/utils'
+require 'puppetserver/utils/http_client'
+require 'puppetserver/utils/file_utilities'
+require 'puppetserver/ca/puppet_config'
+require 'optparse'
+require 'json'
+
+module Puppetserver
+  module Ca
+    class ListAction
+
+      include Puppetserver::Utils
+
+      SUMMARY = 'List all certificate requests'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca list [--help|--version]
+  puppetserver ca list [--config]
+  puppetserver ca list [--all]
+
+Description:
+List outstanding certificate requests. If --all is specified, signed and revoked certificates will be listed as well.
+
+Options:
+      BANNER
+
+      BODY = JSON.dump({desired_state: 'signed'})
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--config CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--all', 'List all certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        return 1 if all_certs.nil?
+
+        requested, signed, revoked = separate_certs(all_certs)
+        input['all'] ? output_certs_by_state(requested, signed, revoked) : output_certs_by_state(requested)
+
+        return 0
+      end
+
+      def output_certs_by_state(requested, signed = [], revoked = [])
+
+        if revoked.empty? && signed.empty? && requested.empty?
+          @logger.inform "No certificates to list"
+          return
+        end
+
+        unless requested.empty?
```

Comment:
i think 
```
if requested.any?
```
is easier to read

Reply:
so @justinstoller had thoughts about why this should be ``empty`` rather than ``any``, but I can't remember exactly what they were... Willing to do either!

Reply:
dunno?  I think it should be fine:
```
[].any?
=> false
irb(main):002:0> [nil].any?
=> false
irb(main):003:0> ["hey"].any?
=> true
```

Reply:
I could go either way. I know some people really don't like `unless`.

Reply:
I think his argument was something about ``if requested.any?`` (with the inverse being  ``unless requested.none?`` and not ``unless requested.empty?``) being syntactically/conceptually different from ``unless empty?`` Is the consensus is that ``if requested.any?`` makes more sense?